### PR TITLE
Performance efficiency improvements for dash::sort

### PR DESCRIPTION
**Pull request by rkowalewski**
This branch has couple of changes to increase efficiency of `dash::sort`.

- Internal functions have been moved to a new header (`internal/Sort-inl.h`)
- There are various const correctness "fixes" to enable more aggressive optimization by the compiler
- Replace 2 collective min / max reductions with a single one as proposed by #562.
- Replace one-sided allreduce with a collective dart reduce.
- Added an all-to-all collective communication in DART since it is needed for efficient data exchange, especially in large-scale use cases. This can be replace with more enhanced versions if #569 is fixed.
- Replace the final local sort with a merge. This drastically increases the performance and may achieve a speedup up to 4 (benchmark results on projekt03) depending on the number of sequences. What we utilize here is the fact that all sequences which have been "received" from other units, are already sorted. We merge these sorted sequences pairwise in a binary tree.

I will provide additional benchmarks results soon.

### CODE REVIEW

File: dart-impl/mpi/src/dart_communication.c
Comment (fmoessbauer) [inline dart-impl/mpi/src/dart_communication.c]: Can we recover from this? It is also not documented in the interface that there is a limit.

File: dash/include/dash/algorithm/Sort.h
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Sort.h]: Wrap this call as it might be not successful

File: dash/include/dash/algorithm/Sort.h
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Sort.h]: Uff... that could be worth a comment ;)

File: dash/include/dash/algorithm/internal/Sort-inl.h
Comment (fmoessbauer) [inline dash/include/dash/algorithm/internal/Sort-inl.h:427]: Probably add a comment?

**Review by fmoessbauer — COMMENTED** [review]
Most of it looks really good. I noticed only some minor points. See my comments inline.

**Comment by rkowalewski** [comment]
Thank you for your good review comments @fmoessbauer. Hopefully everything is addressed now.

File: dash/include/dash/algorithm/Sort.h
Comment (devreal) [inline dash/include/dash/algorithm/Sort.h:233]: Would `valid_partitions` be empty on all units or would the one unit have one partition? If so are  there any collectives on the path taken by that unit? If so we have a problem because the non-empty unit will deadlock...

**Review by devreal — CHANGES_REQUESTED** [review]
Looks good to me, just one thing I noticed...

File: dash/include/dash/algorithm/Sort.h
Comment (rkowalewski) [reply dash/include/dash/algorithm/Sort.h:233]: `valid_partitions` is always equal on all units. ;) So, all units will take this path if no valid partitions have been found.

**Review by devreal — APPROVED** [review]

**Comment by fmoessbauer** [comment]
Why didn't we notice the compilation error in our CI?

**Comment by rkowalewski** [comment]
> Why didn't we notice the compilation error in our CI?

We were wondering about this as well and there is definitely something strange going on. @pascalj is fixing that. ;)

**Comment by fmoessbauer** [comment]
Is this PR ready to merge?

**Review by fmoessbauer — APPROVED** [review]

**Comment by rkowalewski** [comment]
Discussed in the F2F. Ready to merge if unit tests have successfully passed.

**Comment by rkowalewski** [comment]
I am not sure why our codecov often hangs, any ideas @fmoessbauer? It just blocks merging and I do not really prefer the hard "sudo" way to force the merge.

**Comment by fmoessbauer** [comment]
@rkowalewski: There is a problem with external branches (from forks). 

```
==> Uploading reports
    url: https://codecov.io
    query: branch=pull%2F570&commit=4f710f9976e02e755a229a418cd6018ebc8e1d9b&build=4682&build_url=&name=&tag=&slug=dash-project%2Fdash&service=circleci&flags=&pr=570&job=0
    -> Pinging Codecov
HTTP 400
Commit sha does not match Circle build.
```

**Comment by rkowalewski** [comment]
Hm OK. Then I am not using my private repository anymore for pull requests.

**Pull request merged**
